design/open/delta-diff.md

```diff
@@ -0,0 +1,168 @@
+5. UI: GUI only.
+6. Reduce user friction as much as possible.
+
+---
+
+## Non-Goals and off-scope
+
+1.  The Delta diff will be limited to the available Delta Log entries (the JSON files).
+
+---
+
+## High-Level design
+
+### Considered architectures
+
+##### Diff as part of the lakeFS server
+
+Implement the diff functionality within lakeFS.  
+Delta doesn't provide a Go SDK implementation, so we'll need to implement the Delta data model and interface ourselves, which is not something we aim to do.
+In addition, this couples future diff implementations to Go (which might result in similar problems).
+
+##### Diff as a standalone RESTful server
+
+The diff implementation will be implemented in one of the available Delta Lake SDK languages and will be called RESTfully from lakeFS to get the diff results.
+Users will run and manage the server and add communication details to the lakeFS configuration file.  
+The problem with this approach is the friction added for the users. We intend to integrate a diff experience as seamlessly as possible, yet this solution is adding the overhead of managing another server alongside lakeFS.
+
+##### Diff as bundled Rust/WebAssembly in lakeFS
+
+Using the `delta-rs` package to build the differ, and bundling it into the lakeFS binary using `cgo` (to communicate with Rust's FFI) or a Web Assembly package that will use Go as the Web Assembly runtime/engine.  
+Multiple languages can be compiled to WebAssembly which is a great benefit, yet the engines that are needed to run WebAssembly runtime in Go are implemented in other languages (not Go) and are unstable and raise compilation complexity and maintenance.
+
+##### Diff as an external binary executable
+
+We can trigger a diff binary from lakeFS using the `os.exec` package and get the output generated by that binary.
+This is almost where we want to be, except that we'll need to somehow validate that the lakeFS server and the executable are using the same "API version" to communicate, so that the output of the binary would match the expected one from lakeFS. In addition, I would like to decrease the amount of deserialization needed to be implemented to interpret the returned DTO (maintainability and extensibility-wise).
+
+### Chosen architecture
+
+#### The Microkernel/Plugin Architecture
+
+The Microkernel/Plugin architecture is composed of two entities: a single "core system" and multiple "plugins".  
+In our case, the lakeFS server act as the core system, and the different diff implementations, including the Delta diff implementation, will act as plugins.  
+We'll use `gRPC` as the transport protocol, which makes the language of choice almost immaterial (due to the use of protobufs as the transferred data format)
+as long as it's self-contained (theoretically it can also be not system-runtime-dependent but then the cost will be an added requirement for running lakeFS- runtime support).
```

Comment:
> * I need an external binary, built for the correct architecture.
> * I can simplify deployment by bundling the binary inside the lakeFS executable.
> * I need to serialize and deserialize, and can use protobufs if I like.

This is what I meant by "This is almost where we want to be" in the "diff as an external binary executable" section 🙂 

> * I need some way of sharing the schema (JSON schema, Swagger spec, gRPC Protobuf definition, whatever) between the trees.
>  +We will need to figure out how to share protobuf sources between 2 separate trees

This can be solved by adding the plugin code to lakeFS (the same way we add different KV implementations to lakeFS, but possibly using other languages).

The thing I like about the `go-plugin` library is that it's a very lightweight and maintained library that is used in production for quite some time, offers API protocol validation (although a basic one), wraps the gRPC communication, and attaches the logging of the executable for us.

> we will have to write a gRPC server entirely on our own.

True, I actually did this and it's pretty easy. I guess that in both approaches we'll have to implement some kind of entry point to the "plugin". I personally prefer that it would follow some kind of protocol and typed data, but it's arguable (obviously).

Reply:
Long-term we **will** need and have a generic external service interface: think of a user who wants to add or modify a diff/merge service.  I believe they will prefer a looser coupling.
Short- and medium-term I agree with you that using go-plugin makes sense _if_ we believe that it shortens initial development time.

Counterpoint: If we implement the plugin service as a separate REST microservice, can we drive it directly from the GUI, and then we don't need to add anything to lakeFS (server)?  Might skip having to agree with team VE about configuration, adding go-plugin, packaging, etc.